File: src/main/java/ru/octol1ttle/flightassistant/util/events/ChangeLookDirectionEvents.java
package ru.octol1ttle.flightassistant.util.events;

import java.util.Optional;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.Entity;

public final class ChangeLookDirectionEvents {
    private ChangeLookDirectionEvents() {
    }

    public static final Event<Pitch> PITCH = EventFactory.createArrayBacked(
            Pitch.class,
            (listeners) -> (entity, pitchDelta) -> {
                for (Pitch event : listeners) {
                    Optional<Float> result = event.onPitchChange(entity, pitchDelta);
                    if (result.isPresent()) {
                        return result;
                    }
                }
                return Optional.empty();
            }
    );

    @FunctionalInterface
    public interface Pitch {
        Optional<Float> onPitchChange(Entity entity, float pitchDelta);
    }
}

File: src/client/kotlin/ru/octol1ttle/flightassistant/api/event/ChangeLookDirectionEvents.kt
package ru.octol1ttle.flightassistant.api.event

import net.fabricmc.fabric.api.event.*
import net.minecraft.entity.Entity
import ru.octol1ttle.flightassistant.api.event.ChangeLookDirectionEvents.Pitch

class ChangeLookDirectionEvents private constructor() {
    companion object {
        @JvmField
        val PITCH: Event<Pitch> = EventFactory.createArrayBacked(Pitch::class.java)
        { listeners: Array<Pitch> ->
            Pitch { entity, pitchDelta ->
                for (event: Pitch in listeners) {
                    val result: Float? = event.onPitchChange(entity, pitchDelta)
                    if (result != null) {
                        return@Pitch result
                    }
                }
                return@Pitch null
            }
        }
    }

    fun interface Pitch {
        fun onPitchChange(entity: Entity, pitchDelta: Float): Float?
    }
}
